File: lite_agent_sdk/dart/lib/src/agent_message_handler.dart
import 'package:opentool_dart/opentool_dart.dart';
import 'model/model.dart';

abstract class AgentMessageHandler {

  /// When Agent Call Native Function, will callback onFunctionCall
  Future<ToolReturn> onFunctionCall(String sessionId, FunctionCall functionCall) async {
    return ToolReturn(id: functionCall.id, result: FunctionNotSupportedException(functionName: functionCall.name).toJson());
  }

  /// When SSE connection receive a message, will callback onMessage()
  Future<void> onMessage(String sessionId, AgentMessage agentMessage);

  /// When SSE connection receive a chunk, will callback onChunk()
  Future<void> onChunk(String sessionId, AgentMessageChunk agentMessageChunk);

  /// When SSE connection DONE, will callback onDone()
  Future<void> onDone();

  /// When SSE Error, will callback onError(e)
  Future<void> onError(Exception e);

}

File: lite_agent_sdk/csharp/LiteAgentSDK.DotNet/AgentMessageHandler.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using LiteAgentSDK.DotNet.Models;

namespace LiteAgentSDK.DotNet
{
    public abstract class AgentMessageHandler
    {
        private ToolReturn _toolReturn;

        /// <summary>
        /// When the agent calls a native function, this will be called.
        /// </summary>
        public virtual async Task<ToolReturn> OnFunctionCallAsync(string sessionId, FunctionCall functionCall)
        {
            return await Task.Run(() =>
            {
                var result = new Dictionary<string, object>
                {
                    { "error", new FunctionNotSupportedException(functionCall.Name).ToJson() }
                };

                _toolReturn = new ToolReturn(functionCall.Id, result);
                return _toolReturn;
            });
        }

        /// <summary>
        /// When the SSE connection receives a message, this will be called.
        /// </summary>
        public abstract Task OnMessageAsync(string sessionId, AgentMessage agentMessage);

        /// <summary>
        /// When the SSE connection receives a chunk, this will be called.
        /// </summary>
        public abstract Task OnChunkAsync(string sessionId, AgentMessageChunk agentMessageChunk);

        /// <summary>
        /// When the SSE connection is done, this will be called.
        /// </summary>
        public abstract Task OnDoneAsync();

        /// <summary>
        /// When there is an SSE error, this will be called.
        /// </summary>
        public abstract Task OnErrorAsync(Exception e);
    }
}